homeassistant/components/go2rtc/__init__.py

```diff
@@ -54,38 +65,75 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     return True
 
 
 class WebRTCProvider(CameraWebRTCProvider):
     """WebRTC provider."""
 
-    def __init__(self, client: Go2RtcClient) -> None:
+    def __init__(self, hass: HomeAssistant,url:str,) -> None:
         """Initialize the WebRTC provider."""
-        self._client = client
+        self._hass = hass
+        self._url = url
+        self._session = async_get_clientsession(hass)
+        self._rest_client = Go2RtcRestClient(self._session, url)
+        self._sessions: dict[str, Go2RtcWsClient] = {}
 
     async def async_is_supported(self, stream_source: str) -> bool:
         """Return if this provider is supports the Camera as source."""
         return stream_source.partition(":")[0] in _SUPPORTED_STREAMS
 
     async def async_handle_web_rtc_offer(
-        self, camera: Camera, offer_sdp: str
-    ) -> str | None:
-        """Handle the WebRTC offer and return an answer."""
-        streams = await self._client.streams.list()
+        self,
+        camera: Camera,
+        offer_sdp: str,
+        session_id: str,
+        send_message: WebRTCSendMessage,
+    ) -> bool:
+        """Handle the WebRTC offer and return the answer via the provided callback.
+
+        Return value determines if the offer was handled successfully.
+        """
+        self._sessions[session_id] = ws_client = Go2RtcWsClient(
+            self._session, self._url, source=camera.entity_id
+        )
+
+        streams = await self._rest_client.streams.list()
         if camera.entity_id not in streams:
             if not (stream_source := await camera.stream_source()):
-                return None
-            await self._client.streams.add(camera.entity_id, stream_source)
+                return False
+            await self._rest_client.streams.add(camera.entity_id, stream_source)
 
-        answer = await self._client.webrtc.forward_whep_sdp_offer(
-            camera.entity_id, WebRTCSdpOffer(offer_sdp)
-        )
-        return answer.sdp
+        @callback
+        def on_messages(message: ReceiveMessages) -> None:
+            """Handle messages."""
+            value: WebRTCMessages
+            if isinstance(message, WebRTCCandidate):
+                value = HAWebRTCCandidate(message.candidate)
+            elif isinstance(message, WebRTCAnswer):
+                value = HAWebRTCAnswer(message.answer)
+
+            send_message(value)
+
+        ws_client.subscribe(on_messages)
+        await ws_client.send(WebRTCOffer(offer_sdp))
+
+        return True
+
+    async def async_on_webrtc_candidate(self, session_id: str, candidate: str) -> None:
+        """Handle the WebRTC candidate."""
+
+        if ws_client := self._sessions.get(session_id):
+            await ws_client.send(WebRTCCandidate(candidate))
+        else:
+            raise ValueError("Unknown session")
+
+    def close_session(self, session_id: str) -> None:
+        """Close the session."""
+        if ws_client := self._sessions.pop(session_id, None):
```

Comment:
Can this method be called twice that we need to guard for the key not existing ?

Reply:
Right, that's why I was asking. How can it happen that we call `async_close_session` twice? If it would happen, it would be a bug, no? It is tied to the subscription so it should really only be called once in the current code. Better to let it fail in case that assumption is wrong, as it might hide another bug when that happens.